roachprod: add VM label for "spot" instance

nameisbhaskar opened the pull request:
what was there before: Previously, there was no way to identify spot instances using a label for tracking (billing)
why it needed to change: This was inadequate because the VMs can be tracked for billing using the added label
what you did about it: To address this, this patch adds a new tag as "spot:true"

Fixes: #119629
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119763)
<!-- Reviewable:end -->

BabuSrithar commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -26,6 +27,24 @@ import (
+
+// printDetailsColumnHeaders are the headers to be printed in the defined sequence.
+var printDetailsColumnHeaders = []string{
+	Name, DNS, PrivateIP, PublicIP, MachineType, CPUArch, CPUFamily, IsSpot,
```
I think we can name this column "Provision Model" instead of  "IsSpot"

herkolategan reviewed (changes requested):
Nice work! A few comments and things to address:

I noticed the multiple commits in this PR. We amend commits rather than adding commits to address a change that is part of an existing commit. See: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/2647261318/Updating+PRs+during+review for details.
All commits should address a package as well.

The commit message only needs to contain the issue, you can omit the jira link CRDB-36220 (unless specifying an epic). You can also move the "Epic: None" to the bottom after "fixes", just for clarity.

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -155,9 +179,43 @@ func (c *Cluster) PrintDetails(logger *logger.Logger) {
+func logPrettifiedHeader(logger *logger.Logger, headers []string) {
+	row := ""
+	for _, header := range headers {
+		row = fmt.Sprintf("%s%s\t", row, header)
```
Nit: I've run `roachprod list` by itself (without details) and `tabwriter.NewWriter` seems to do a better job of aligning the columns [1]. On my system the current print out is misaligned.
[1] https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachprod/main.go#L322

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -155,9 +179,43 @@ func (c *Cluster) PrintDetails(logger *logger.Logger) {
+		row = fmt.Sprintf("%s%s\t", row, header)
+	}
+	separator := strings.Repeat("-", len(row))
+	logger.Printf(separator)
```
Nit: I would leave out these rows, especially if using `tabwriter` since they tend to just take up real estate. There is already a split by cluster when the cluster name is printed out, which is enough separation.

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -26,6 +27,29 @@ import (
 
+const (
+	// The following constants are headers that are used for printing the VM details.
+
```
Nit: This space/return seems out of place.

herkolategan reviewed (commented):
Thanks for making the changes. I've played around a bit with the new print details locally, and left some minor comments regarding the format.

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -155,9 +179,46 @@ func (c *Cluster) PrintDetails(logger *logger.Logger) {
+			headerProvisionModel: provisionModel,
+		}))
+	}
+	logger.Errorf("Failed to print details for VMs: %v", fmt.Errorf("hgello"))
```
Thinks this slipped in accidentally.

herkolategan replied:
```
09:36:00 cluster_cloud.go:171: teamcity-14227344-1709533889-36-n4cpu16: [aws]
09:36:00 cluster_cloud.go:177: 11h24m0s remaining
09:36:00 cluster_cloud.go:197: Failed to print details for VMs: hgello
```

herkolategan commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -155,9 +179,46 @@ func (c *Cluster) PrintDetails(logger *logger.Logger) {
+	}
+	logger.Errorf("Failed to print details for VMs: %v", fmt.Errorf("hgello"))
+	if err := tw.Flush(); err != nil {
+		logger.Errorf("Failed to print details for VMs: %v", err)
```
May be better to bubble this error up and have `PrintDetails` return an error, which will have the loop running `PrintDetails` check for the error and return from there as wel..

nameisbhaskar replied:
Sure! I was just not sure whether a print failure should fail the flow. Will add the error return!

nameisbhaskar committed:
roachprod: add VM label for "spot" instance

what was there before: Previously, there was no way to identify spot instances using a label for tracking (billing)
why it needed to change: This was inadequate because the VMs can be tracked for billing using the added label
what you did about it: To address this, this patch adds a new tag as "spot:true"

Fixes: #119629
Epic: none

herkolategan reviewed (approved):
Thanks for making the changes!

nameisbhaskar commented:
bors r+

srosenberg commented:
blathers backport 23.1 23.2

nameisbhaskar commented:
Thanks @herkolategan and @BabuSrithar for reviewing the changes!

renatolabs commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -155,9 +179,43 @@ func (c *Cluster) PrintDetails(logger *logger.Logger) {
 		logger.Printf("(no expiration)")
 	}
+	// Align columns left and separate with at least two spaces.
+	tw := tabwriter.NewWriter(os.Stdout, 0, 8, 2, ' ', 0)
```
I noticed that we used to write this cluster information on the `logger` passed to this function, but now we're writing straight to stdout.

This means that this information is now printed on TeamCity's "Build Log" instead of the corresponding cluster creation log as before (where I think it's more useful).

Can we revert back to writing this info to the logger?

nameisbhaskar replied:
Thanks @herkolategan for creating https://github.com/cockroachdb/cockroach/pull/120118
